### ISSUE TITLE
[Bug]: [tptr lowering] TokenEmbedding backward fails on dynamic tl.atomic_add in embedding_dense_backward

### ISSUE THREAD

**Issue by 111a909** [RuyiAI-Stack/triton-riscv#38]
### Triton python code

```mlir
import torch
import triton
import triton.language as tl


@triton.jit
def _embedding_dense_backward_kernel(
    grad_output_ptr,
    indices_ptr,
    grad_weight_ptr,
    num_weights,
    padding_idx,
    BLOCK_D: tl.constexpr,
    EMBED_DIM: tl.constexpr,
):
    pid_n = tl.program_id(0)
    pid_d = tl.program_id(1)

    offs_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
    mask_d = offs_d < EMBED_DIM

    idx = tl.load(indices_ptr + pid_n)
    valid = (idx != padding_idx) & (idx >= 0) & (idx < num_weights)

    go_ptrs = grad_output_ptr + pid_n * EMBED_DIM + offs_d
    go = tl.load(go_ptrs, mask=mask_d, other=0).to(tl.float32)

    gw_ptrs = grad_weight_ptr + idx * EMBED_DIM + offs_d
    mask = mask_d & valid
    tl.atomic_add(gw_ptrs, go, mask=mask)


@triton.jit
def _embedding_dense_backward_count_kernel(
    indices_ptr,
    counts_ptr,
    N,
    num_weights,
    padding_idx,
    BLOCK_N: tl.constexpr,
):
    pid = tl.program_id(0)
    offs = pid * BLOCK_N + tl.arange(0, BLOCK_N)
    mask = offs < N
    idx = tl.load(indices_ptr + offs, mask=mask, other=0).to(tl.int32)
    valid = mask & (idx != padding_idx) & (idx >= 0) & (idx < num_weights)
    tl.atomic_add(counts_ptr + idx, 1, mask=valid)


@triton.jit
def _embedding_dense_backward_kernel_scale_by_freq(
    grad_output_ptr,
    indices_ptr,
    counts_ptr,
    grad_weight_ptr,
    num_weights,
    padding_idx,
    BLOCK_D: tl.constexpr,
    EMBED_DIM: tl.constexpr,
):
    pid_n = tl.program_id(0)
    pid_d = tl.program_id(1)

    offs_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
    mask_d = offs_d < EMBED_DIM

    idx = tl.load(indices_ptr + pid_n).to(tl.int32)
    valid = (idx != padding_idx) & (idx >= 0) & (idx < num_weights)

    go_ptrs = grad_output_ptr + pid_n * EMBED_DIM + offs_d
    go = tl.load(go_ptrs, mask=mask_d, other=0.0)

    cnt = tl.load(counts_ptr + idx, mask=valid, other=1)
    go = go / cnt

    gw_ptrs = grad_weight_ptr + idx * EMBED_DIM + offs_d
    mask = mask_d & valid
    tl.atomic_add(gw_ptrs, go, mask=mask)


def embedding_dense_backward(
    grad_output: torch.Tensor,
    indices: torch.Tensor,
    num_weights: int,
    padding_idx: int,
    scale_grad_by_freq: bool,
):
    assert indices.dtype in (
        torch.int32,
        torch.int64,
    ), "Indices must be int32 or int64."

    device = grad_output.device
    assert grad_output.dim() >= 2, (
        "grad_output must have embedding dimension as the last dim."
    )

    D = grad_output.shape[-1]
    go = grad_output.contiguous().view(-1, D)  # (N, D)
    idx = indices.contiguous().view(-1)
    N = idx.numel()

    assert go.shape[0] == N, "indices number must match grad_output rows."
    grad_weight_fp32 = torch.zeros(
        (num_weights, D), device=device, dtype=torch.float32
    )

    BLOCK_D = 128
    grid = (N, triton.cdiv(D, BLOCK_D))

    if scale_grad_by_freq:
        counts = torch.zeros((num_weights,), device=device, dtype=torch.int32)
        BLOCK_N = 512
        _embedding_dense_backward_count_kernel[(triton.cdiv(N, BLOCK_N),)](
            idx,
            counts,
            N,
            num_weights,
            padding_idx if padding_idx is not None else -1,
            BLOCK_N=BLOCK_N,
        )

        _embedding_dense_backward_kernel_scale_by_freq[grid](
            go,
            idx,
            counts,
            grad_weight_fp32,
            num_weights,
            padding_idx if padding_idx is not None else -1,
            BLOCK_D=BLOCK_D,
            EMBED_DIM=D,
        )
    else:
        _embedding_dense_backward_kernel[grid](
            go,
            idx,
            grad_weight_fp32,
            num_weights,
            padding_idx if padding_idx is not None else -1,
            BLOCK_D=BLOCK_D,
            EMBED_DIM=D,
        )

    if grad_output.dtype != torch.float32:
        return grad_weight_fp32.to(grad_output.dtype)
    return grad_weight_fp32
```

### Triton IR

```mlir
#loc = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":12:1)
#loc20 = loc("grad_output_ptr"(#loc))
#loc21 = loc("indices_ptr"(#loc))
#loc22 = loc("grad_weight_ptr"(#loc))
#loc23 = loc("num_weights"(#loc))
#loc24 = loc("padding_idx"(#loc))
module {
  tt.func public @_embedding_dense_backward_kernel(%grad_output_ptr: !tt.ptr<f32> loc("grad_output_ptr"(#loc)), %indices_ptr: !tt.ptr<i64> loc("indices_ptr"(#loc)), %grad_weight_ptr: !tt.ptr<f32> loc("grad_weight_ptr"(#loc)), %num_weights: i32 loc("num_weights"(#loc)), %padding_idx: i32 loc("padding_idx"(#loc))) attributes {noinline = false} {
    %go = arith.constant dense<0.000000e+00> : tensor<16xf32> loc(#loc25)
    %valid = arith.constant 0 : i64 loc(#loc26)
    %gw_ptrs = arith.constant 16 : i64 loc(#loc27)
    %mask_d = arith.constant dense<16> : tensor<16xi32> loc(#loc28)
    %c16_i32 = arith.constant 16 : i32 loc(#loc5)
    %pid_n = tt.get_program_id x : i32 loc(#loc29)
    %pid_d = tt.get_program_id y : i32 loc(#loc30)
    %offs_d = arith.muli %pid_d, %c16_i32 : i32 loc(#loc31)
    %offs_d_0 = tt.make_range {end = 16 : i32, start = 0 : i32} : tensor<16xi32> loc(#loc32)
    %offs_d_1 = tt.splat %offs_d : i32 -> tensor<16xi32> loc(#loc31)
    %offs_d_2 = arith.addi %offs_d_1, %offs_d_0 : tensor<16xi32> loc(#loc31)
    %mask_d_3 = arith.cmpi slt, %offs_d_2, %mask_d : tensor<16xi32> loc(#loc28)
    %idx = tt.addptr %indices_ptr, %pid_n : !tt.ptr<i64>, i32 loc(#loc33)
    %idx_4 = tt.load %idx : !tt.ptr<i64> loc(#loc34)
    %valid_5 = arith.extsi %padding_idx : i32 to i64 loc(#loc35)
    %valid_6 = arith.cmpi ne, %idx_4, %valid_5 : i64 loc(#loc35)
    %valid_7 = arith.cmpi sge, %idx_4, %valid : i64 loc(#loc26)
    %valid_8 = arith.andi %valid_6, %valid_7 : i1 loc(#loc36)
    %valid_9 = arith.extsi %num_weights : i32 to i64 loc(#loc37)
    %valid_10 = arith.cmpi slt, %idx_4, %valid_9 : i64 loc(#loc37)
    %valid_11 = arith.andi %valid_8, %valid_10 : i1 loc(#loc36)
    %go_ptrs = arith.muli %pid_n, %c16_i32 : i32 loc(#loc38)
    %go_ptrs_12 = tt.addptr %grad_output_ptr, %go_ptrs : !tt.ptr<f32>, i32 loc(#loc39)
    %go_ptrs_13 = tt.splat %go_ptrs_12 : !tt.ptr<f32> -> tensor<16x!tt.ptr<f32>> loc(#loc39)
    %go_ptrs_14 = tt.addptr %go_ptrs_13, %offs_d_2 : tensor<16x!tt.ptr<f32>>, tensor<16xi32> loc(#loc39)
    %go_15 = tt.load %go_ptrs_14, %mask_d_3, %go : tensor<16x!tt.ptr<f32>> loc(#loc25)
    %gw_ptrs_16 = arith.muli %idx_4, %gw_ptrs : i64 loc(#loc27)
    %gw_ptrs_17 = tt.addptr %grad_weight_ptr, %gw_ptrs_16 : !tt.ptr<f32>, i64 loc(#loc40)
    %gw_ptrs_18 = tt.splat %gw_ptrs_17 : !tt.ptr<f32> -> tensor<16x!tt.ptr<f32>> loc(#loc40)
    %gw_ptrs_19 = tt.addptr %gw_ptrs_18, %offs_d_2 : tensor<16x!tt.ptr<f32>>, tensor<16xi32> loc(#loc40)
    %mask = tt.splat %valid_11 : i1 -> tensor<16xi1> loc(#loc41)
    %mask_20 = arith.andi %mask_d_3, %mask : tensor<16xi1> loc(#loc41)
    %0 = tt.atomic_rmw fadd, acq_rel, gpu, %gw_ptrs_19, %go_15, %mask_20 : (tensor<16x!tt.ptr<f32>>, tensor<16xf32>, tensor<16xi1>) -> tensor<16xf32> loc(#loc19)
    tt.return loc(#loc)
  } loc(#loc)
} loc(#loc)
#loc1 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":31:10)
#loc2 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":28:37)
#loc3 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":33:33)
#loc4 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":25:14)
#loc5 = loc(unknown)
#loc6 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":21:13)
#loc7 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":22:13)
#loc8 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":24:14)
#loc9 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":24:32)
#loc10 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":27:19)
#loc11 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":27:11)
#loc12 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":28:14)
#loc13 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":28:13)
#loc14 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":28:50)
#loc15 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":30:33)
#loc16 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":30:15)
#loc17 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":33:15)
#loc18 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":34:12)
#loc19 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py":35:5)
#loc25 = loc("go"(#loc1))
#loc26 = loc("valid"(#loc2))
#loc27 = loc("gw_ptrs"(#loc3))
#loc28 = loc("mask_d"(#loc4))
#loc29 = loc("pid_n"(#loc6))
#loc30 = loc("pid_d"(#loc7))
#loc31 = loc("offs_d"(#loc8))
#loc32 = loc("offs_d"(#loc9))
#loc33 = loc("idx"(#loc10))
#loc34 = loc("idx"(#loc11))
#loc35 = loc("valid"(#loc12))
#loc36 = loc("valid"(#loc13))
#loc37 = loc("valid"(#loc14))
#loc38 = loc("go_ptrs"(#loc15))
#loc39 = loc("go_ptrs"(#loc16))
#loc40 = loc("gw_ptrs"(#loc17))
#loc41 = loc("mask"(#loc18))
```

### Crash log

```shell
When testing python/examples/flaggems/embedding_dense_backward.py, lowering fails in indice_freq_kernel and embedding_backward_kernel.

Observed error:

error: unexpected op in ptr sequence
warning: Cannot transform tensor of pointers into a single base pointer with tensor of offsets
error: Dialect `tptr` not found for custom op `tptr.type_offset`

The failure appears when dynamic-address tl.atomic_add is lowered.
```

## Problem

When testing `python/examples/flaggems/embedding_dense_backward.py`, lowering fails for dynamic-address `tl.atomic_add` in:

- `indice_freq_kernel`
- `embedding_backward_kernel`

The failure seems related to tensor-of-pointers / tptr lowering.

## Error

```text
error: unexpected op in ptr sequence
warning: Cannot transform tensor of pointers into a single base pointer with tensor of offsets
error: Dialect `tptr` not found for custom op `tptr.type_offset`

**Comment by maochuanjun** [comment]
Thanks for the detailed report. This issue has been fixed by #44 and has been verified against the latest `main`.

  The root cause was in the unstructured pointer/atomic lowering path:

  - `TritonToUnstructured` did not recognize `tt.atomic_rmw` as a supported consumer of a dynamically derived tensor of pointers.
  - Therefore, a pointer such as `grad_weight_ptr + idx * EMBED_DIM + offs_d` reached the fallback case and produced `unexpected op in ptr sequence`.
  - The unstructured pointer conversion was then left incomplete, allowing `tptr.type_offset` to remain in the IR passed to the next compilation stage.
  - The subsequent `Dialect 'tptr' not found` error was therefore a secondary symptom of the failed atomic pointer normalization, rather than an independent operator issue.

  PR #44 adds the complete lowering path for dynamic-address atomic operations:

  ```text
  tt.atomic_rmw
    -> tts.atomic_rmw (base pointer + dynamic offsets)
    -> memref.atomic_rmw
    -> llvm.atomicrmw
  ```

  The mask, atomic operation and memory scope are preserved during the conversion. Both floating-point accumulation and the integer count accumulation used by embedding_dense_backward are covered.

  This covers:
  - scale_grad_by_freq=False and True;
  - padding_idx=None, 0, and 5;
  - 1-D, 2-D, and 3-D index shapes;
  - embedding dimensions 32, 64, and 128.

  This issue should be closed as fixed. If there is anything still wrong, please provide a detailed explanation.
